vpr/src/base/clustered_netlist.h

```diff
@@ -148,9 +148,13 @@ class ClusteredNetlist : public Netlist<ClusterBlockId, ClusterPortId, ClusterPi
      */
 
     //Returns the physical pin index (i.e. pin index on the
-    //t_logical_block_type) of the specified logical pin
+    //t_physical_tile_type) of the specified logical pin
     int pin_physical_index(const ClusterPinId id) const;
 
+    //Returns the logical pin index (i.e. pin index on the
+    //t_logical_block_type) of the cluster pin
+    int pin_logical_index(const ClusterPinId pin_id) const;
```

Comment:
Do we need to store both these pieces of information in the netlist?

While knowing the pin index within the logical block type makes sense to have in the netlist, I'm less certain about the pin_physical_index (i.e. the pin index of the tile where the block is placed). That information is really dependent on placement and is not directly clustered netlist information.

It seems like pin_physical_index() would be better as a free function (i.e. not a member of ClusteredNetlist), which looked up this information directly from the placement & tile to logical pin mapping. So something like
```cpp
int pin_tile_index(const ClusterPinId pin) {
    //Get the logical pin index of pin within it's logical block type

   //Determine the physical tile type implementing the logical block (i.e. based on the current placement)

   //Use the physical/logical pin mapping to convert the logical block pin index to the physical tile pin index & return it
}
```

This would avoid duplicating this state information in the CLB netlist and avoid a whole class of potential bugs, where due to a placement change the physical tile pin index stored in the ClusteredNetlist gets out of date.

Reply:
I believe that, from a theoretical point of view you are completely right, the pin physical index should not be part of the clustered netlist, as it shouldn't be aware of the location chosen for a specific clb, hence its physical pins. 

From what I have learned, this information (physical pin related to a `ClusterPinId`) is stored in a fast-accessible structure. If we extract this in an external function, I believe that run-time will be affected. This because for each `ClusteredPinId` we would need to get many different information (adopted physical tile, logical pin, and physical pin from the physical/logical pin mapping).

I may be wrong and run-time will not be affected at all, therefore I will try to get rid of both `pin_physical_index`, `net_pin_physical_index` and `set_pin_physical_index` and extract them from the clustered netlist.

Reply:
Let me rephrase. Actually I think that after all this is doable as a lot of `pin_physical_index` can be swapped to `pin_logical_index`. Pin physical index will be needed only after placement only and there are not many places where it is used, so run-time should not be affected by this.

Reply:
As predicted repeating the lookup on the direct map several times for each pin increases run-time of a non-negligible amount.

I have opted to store the `ClusterPinId <-> physical_pin_mapping` in the placement context. The mapping gets computed after the initial placement and modified at the end of the whole placement process, taking into account also the `z` location in tiles having `capacity > 1`.

I think placement context is a better place to store this kind of information.

Reply:
Yes that seems like a better place for it!